File: src/main/java/me/Salt/Logging/LogUtils.java
/*
 * Copyright (c) 2017 DevJake
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.Salt.Logging;

import me.Salt.Main;
import me.Salt.Util.Language.LangString;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class LogUtils {
    private List<LogEntry> logEntries = new ArrayList<>();
    
    private static boolean notSafe() {
        return Main.salt == null ? false : Main.salt.getLanguageHandler() == null ? false : true;
    }
    
    public static void log(String message, Severity severity, String className, Exception e) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(severity) + "] " + message);
    }
    
    public static void log(String message, String className, Exception e) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(Severity.INFO) + "] " + message);
    }
    
    public static void log(String message, String className) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(Severity.INFO) + "] " + message);
    }
    
    public static void log(String message, Severity severity, String className) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(severity) + "] " + message);
    }
    
    public static void log(String message, Severity severity) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(severity) + "] " + message);
    }
    
    public static void log(String message) {
        if (notSafe()) return;
        System.out.println("[" + Instant.now().toString() + " ][" + getLang(Severity.INFO) + "] " + message);
    }
    //TODO consider adding a Log builder, to properly format a log entry
    
    private static String getLang(Severity severity) {
        switch (severity) {
            case INFO:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_INFO) != null ? Main.salt.getLanguageHandler()
                                                                                       .getLanguage(
                                                                                               Main.salt.getDefaultLangCode())
                                                                                       .getString(
                                                                                               LangString.LOGGING_INFO)
                                                                                       .toUpperCase() : "INFO";
            case DEBUG:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_DEBUG) != null ? Main.salt.getLanguageHandler()
                                                                                        .getLanguage(
                                                                                                Main.salt.getDefaultLangCode())
                                                                                        .getString(
                                                                                                LangString.LOGGING_DEBUG)
                                                                                        .toUpperCase() : "DEBUG";
            case FATAL:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_FATAL) != null ? Main.salt.getLanguageHandler()
                                                                                        .getLanguage(
                                                                                                Main.salt.getDefaultLangCode())
                                                                                        .getString(
                                                                                                LangString.LOGGING_FATAL)
                                                                                        .toUpperCase() : "FATAL";
            case CONFIG:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_CONFIG) != null ? Main.salt.getLanguageHandler()
                                                                                         .getLanguage(
                                                                                                 Main.salt.getDefaultLangCode())
                                                                                         .getString(
                                                                                                 LangString.LOGGING_CONFIG)
                                                                                         .toUpperCase() : "CONFIG";
            case SEVERE:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_SEVERE) != null ? Main.salt.getLanguageHandler()
                                                                                         .getLanguage(
                                                                                                 Main.salt.getDefaultLangCode())
                                                                                         .getString(
                                                                                                 LangString.LOGGING_SEVERE)
                                                                                         .toUpperCase() : "SEVERE";
            case WARNING:
                return Main.salt.getLanguageHandler()
                                .getLanguage(Main.salt.getDefaultLangCode())
                                .getString(LangString.LOGGING_WARNING) != null ? Main.salt.getLanguageHandler()
                                                                                          .getLanguage(
                                                                                                  Main.salt.getDefaultLangCode())
                                                                                          .getString(
                                                                                                  LangString.LOGGING_WARNING)
                                                                                          .toUpperCase() : "WARNING";
            default:
                return "LANGUAGE FILE MISSING DATA";
        }
    }
    
    public List<LogEntry> getLogEntries() {
        return logEntries;
    }
    
    public enum Severity {
        INFO,
        CONFIG,
        WARNING,
        SEVERE,
        FATAL,
        DEBUG,
    }
}

File: src/main/kotlin/me/salt/logging/LogUtils.kt
/*
 * Copyright 2017 DevJake
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.salt.logging

import me.salt.config.Configs
import me.salt.config.entities.SaltLogConfig
import me.salt.events.Event
import me.salt.exception.ConfigMissingValueException
import me.salt.exception.exception
import me.salt.objects.getConfig
import java.time.Instant

object LogUtils {
    private val cache = mutableListOf<LogEntry>()
    private fun addLogEntry(entry: LogEntry) {
        cache.add(entry)
        flush(entry)
    }

    private fun flushCache() {
        if (cache.size < 100) return

        cache.forEach { entry ->
            println(entry.toString())
        }

        cache.clear()

//TODO("Writing entries (with log entry about the flush), clearing cache"), splitting logs to different files, such as one for exceptions
    }

    private fun flush(entry: LogEntry) {
        var logConfig = Configs.salt.LOG_CONFIG.getConfig(SaltLogConfig::class.java) //TODO
        if (logConfig == null)
            exception(ConfigMissingValueException())
        if (logConfig?.logToConsole as Boolean) println(calcLogMessage(entry))

        flushCache()
    }

    private fun calcLogMessage(entry: LogEntry) =
            "[${entry.dateTime}] " +
                    "[${entry.type.entryName}" +
                    "${if (entry.optional != null)
                        "/${entry.optional}" else ""}] " +
                    entry.elements.joinToString()

    fun info(optional: String? = null) =
            Logger(LogType.INFO, optional)

    fun debug(optional: String? = null) =
            Logger(LogType.DEBUG, optional)

    fun warn(optional: String? = null) =
            Logger(LogType.WARNING, optional)

    fun fatal(optional: String? = null) =
            Logger(LogType.FATAL, optional)

    fun severe(optional: String? = null) =
            Logger(LogType.SEVERE, optional)

    class Logger(private val type: LogType, private val optional: String?) {
        fun log(message: String) =
                addLogEntry(LogEntry(type, optional, message))

        fun log(exception: Exception) =
                addLogEntry(LogEntry(type, optional, exception))

        fun log(message: String, exception: Exception) =
                addLogEntry(LogEntry(type, optional, message, exception))

        fun log(event: Event) =
                addLogEntry(LogEntry(type, optional, event))

        fun log(message: String, event: Event) =
                addLogEntry(LogEntry(type, optional, message, event))

    }

    enum class LogType(var entryName: String) {
        INFO("INFO"),
        DEBUG("DEBUG"),
        WARNING("WARN"),
        SEVERE("SEVERE"),
        FATAL("FATAL");

        init {
            entryName = entryName.toUpperCase()
        }
    }

    private class LogEntry(val type: LogType, val optional: String?, vararg val elements: Any) {
        val dateTime = Instant.now()
        val saltId: String = "tempSaltId" //TODO Add SaltId generator
        //TODO generate the log message from the provided element(s) and other information.
        //TODO switch through types of loggable entities to determine best formatting

    }
}

fun logInfo(message: String, optional: String? = null) = LogUtils.info(optional).log(message)
fun logDebug(message: String, optional: String? = null) = LogUtils.debug(optional).log(message)
fun logWarn(message: String, optional: String? = null) = LogUtils.warn(optional).log(message)
fun logSevere(message: String, optional: String? = null) = LogUtils.severe(optional).log(message)
fun logFatal(message: String, optional: String? = null) = LogUtils.fatal(optional).log(message)

fun logEasy(message: String) = LogUtils.info(Configs.salt.LOG_CONFIG.getConfig(SaltLogConfig::class.java)?.easyLogMessage ?: "EASYLOG").log(message)